File: src/core/ports/cross_chain_port.py
from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Dict, List, Optional

from src.shared.types import Address, Chain, CrossChainMessage, Hash, HexString, WeiAmount


class ICrossChainBridgePort(ABC):
    @abstractmethod
    async def lock_assets(
        self,
        source_chain: Chain,
        target_chain: Chain,
        source_address: Address,
        target_address: Address,
        amount: WeiAmount,
        token_address: Optional[Address] = None,
        data: HexString = "0x",
    ) -> CrossChainMessage: ...

    @abstractmethod
    async def mint_assets(
        self,
        message: CrossChainMessage,
        proof: HexString,
    ) -> Hash: ...

    @abstractmethod
    async def verify_message(
        self,
        message: CrossChainMessage,
        proof: HexString,
    ) -> bool: ...

    @abstractmethod
    async def get_message_status(
        self,
        message_id: str,
    ) -> CrossChainMessage: ...

    @abstractmethod
    async def list_messages(
        self,
        source_chain: Optional[Chain] = None,
        target_chain: Optional[Chain] = None,
        status: Optional[str] = None,
        limit: int = 100,
    ) -> List[CrossChainMessage]: ...

    @abstractmethod
    async def retry_message(
        self,
        message_id: str,
    ) -> bool: ...

    @abstractmethod
    async def cancel_message(
        self,
        message_id: str,
    ) -> bool: ...

    @abstractmethod
    async def generate_proof(
        self,
        message: CrossChainMessage,
        source_tx_hash: Hash,
    ) -> HexString: ...


class IMessageValidator(ABC):
    @abstractmethod
    async def validate_source_transaction(
        self,
        source_chain: Chain,
        tx_hash: Hash,
        message: CrossChainMessage,
    ) -> bool: ...

    @abstractmethod
    async def validate_signature(
        self,
        message: CrossChainMessage,
        signature: HexString,
        signer_address: Address,
    ) -> bool: ...

    @abstractmethod
    async def validate_nonce(
        self,
        message: CrossChainMessage,
    ) -> bool: ...


class IAtomicExecutor(ABC):
    @abstractmethod
    async def execute_atomic(
        self,
        operations: List[Dict[str, Any]],
        timeout: int = 300,
    ) -> bool: ...

    @abstractmethod
    async def rollback(
        self,
        operation_id: str,
    ) -> bool: ...

    @abstractmethod
    async def prepare(
        self,
        operations: List[Dict[str, Any]],
    ) -> str: ...

    @abstractmethod
    async def commit(
        self,
        operation_id: str,
    ) -> bool: ...

File: src/core/ports/crossChain.port.ts
import type { ChainId, Address, Hash, HexString, WeiAmount, UUID } from '@shared/types';
import type { CrossChainMessage } from '@core/domain/blockchain';

export interface CrossChainProof {
  messageHash: Hash;
  sourceChainId: ChainId;
  targetChainId: ChainId;
  proof: HexString[];
  blockHash: Hash;
  transactionHash: Hash;
  timestamp: number;
}

export interface MessageVerificationResult {
  isValid: boolean;
  message?: CrossChainMessage;
  error?: string;
  confirmations: number;
  requiredConfirmations: number;
}

export interface AtomicOperation {
  id: UUID;
  sourceOperation: {
    chainId: ChainId;
    type: 'lock' | 'burn';
    transactionHash?: Hash;
    address: Address;
    amount: WeiAmount;
  };
  targetOperation: {
    chainId: ChainId;
    type: 'mint' | 'unlock';
    transactionHash?: Hash;
    address: Address;
    amount: WeiAmount;
  };
  status: 'pending' | 'source_confirmed' | 'source_executed' | 'target_executing' | 'completed' | 'failed' | 'rolled_back';
  createdAt: number;
  completedAt?: number;
}

export interface CrossChainBridgePort {
  initiateBridge(
    sourceChainId: ChainId,
    targetChainId: ChainId,
    sourceAddress: Address,
    targetAddress: Address,
    amount: WeiAmount,
    data?: HexString
  ): Promise<CrossChainMessage>;

  verifyMessage(
    messageHash: Hash,
    sourceChainId: ChainId,
    targetChainId: ChainId,
    requiredConfirmations?: number
  ): Promise<MessageVerificationResult>;

  generateProof(messageHash: Hash, sourceChainId: ChainId): Promise<CrossChainProof>;

  executeBridge(
    proof: CrossChainProof,
    targetSigner?: Address
  ): Promise<{
    transactionHash: Hash;
    messageId: UUID;
  }>;

  getMessage(messageHash: Hash): Promise<CrossChainMessage | null>;

  getAtomicOperation(operationId: UUID): Promise<AtomicOperation | null>;

  initiateAtomicBridge(
    sourceChainId: ChainId,
    targetChainId: ChainId,
    sourceAddress: Address,
    targetAddress: Address,
    amount: WeiAmount,
    data?: HexString
  ): Promise<AtomicOperation>;

  confirmSourceOperation(
    operationId: UUID,
    transactionHash: Hash
  ): Promise<AtomicOperation>;

  rollbackOperation(
    operationId: UUID,
    reason?: string
  ): Promise<AtomicOperation>;
}

export interface BridgeValidatorPort {
  validateLockTransaction(
    chainId: ChainId,
    transactionHash: Hash,
    expectedAmount: WeiAmount,
    expectedRecipient: Address
  ): Promise<boolean>;

  validateMintTransaction(
    chainId: ChainId,
    transactionHash: Hash,
    expectedAmount: WeiAmount,
    expectedRecipient: Address
  ): Promise<boolean>;

  verifyMessageIntegrity(message: CrossChainMessage, signature: HexString): Promise<boolean>;
}
